kube-runtime/src/watcher.rs

```diff
@@ -48,28 +48,24 @@ pub enum Event<K> {
     Delete(K),
     /// The watch stream was restarted.
     ///
-    /// If using the `ListWatch` strategy, `RestartPage` events will follow this event.
-    /// If using the `StreamingList` strategy, this event will be followed by `RestartApply` or `RestartDelete` events.
-    RestartInit,
-    /// A page of objects was received during the restart.
+    /// If using the `ListWatch` strategy, this is a relist, and indicates that `InitlPage` events will follow.
+    /// If using the `StreamingList` strategy, this event will be followed by `InitApply` + `InitDelete` events.
+    Init,
+    /// A page of objects was received during `Init` with the `ListWatch` strategy.
     ///
-    /// This event is only returned when using the `ListWatch` strategy.
+    /// Any objects that were previously [`Applied`](Event::Applied) but are not listed in any of
+    /// the `InitPage` events should be assumed to have been [`Deleted`](Event::Deleted).
+    InitPage(Vec<K>),
+    /// An object received during `Init` with the `StreamingList` strategy.
     ///
-    /// Any objects that were previously [`Applied`](Event::Applied) but are not listed in any of the pages
-    /// should be assumed to have been [`Deleted`](Event::Deleted).
-    RestartPage(Vec<K>),
-    /// An object was added or modified during the initial watch.
-    ///
-    /// This event is only returned when using the `StreamingList` strategy.
-    RestartApply(K),
-    /// An object was deleted during the initial watch.
-    ///
-    /// This event is only returned when using the `StreamingList` strategy.
-    RestartDelete(K),
-    /// The restart is complete.
+    /// Any objects that were previously [`Applied`](Event::Applied) but are not listed in any of
+    /// the `InitAdd` events should be assumed to have been [`Deleted`](Event::Deleted).
+    InitApply(K),
```

Comment:
These could be merged into one `InitPage(SmallVec<[K; 1]>)`, I think? (Using https://docs.rs/smallvec/latest/smallvec/struct.SmallVec.html)

That way, only users that actually care about the distinction have to care, while you still avoid the allocation overhead if it isn't needed.

Reply:
Another option would be to extract the pagination strategy to a trait/generic.

Reply:
That should be possible yes. My only worry is that `InitPage` can go away (in like 3 years) when `StreamingList` is stabilised and covered in our MK8SV, at which point it's not necessary to have a vector there at all.

The vector part on the receiver is only slightly more verbose though https://github.com/kube-rs/kube/blob/59efd8de8087edc206c8a0cad9249fe87d9d271a/kube-runtime/src/reflector/store.rs#L114-L125

Reply:
I think extracting the strategy into a trait somehow could be useful for many reasons though! But would like to iron out deficiencies in shared streams controllers firsts.

Reply:
You shouldn't need to collect into the `AHashMap` fwiw, `Extend::extend` takes an iterator rather than a concrete collection. So the smallveced version would end up as "just":

```
 watcher::Event::InitPage(new_objs) => { 
     let new_objs = new_objs 
         .iter() 
         .map(|obj| (obj.to_object_ref(self.dyntype.clone()), Arc::new(obj.clone()))) ;
     self.buffer.extend(new_objs); 
 }
```

Alternatively, if we want to lean into `InitApply`, we could also _always_ flatten the list (so emit multiple `InitApply` when we receive an `InitPage` event).

Reply:
TIL. Extend trick applied: https://github.com/kube-rs/kube/pull/1499/commits/4f69a37435484699f53f235156ccede860f1b690 :+1:

Reply:
> Alternatively, if we want to lean into InitApply as the path forward, we could also always flatten the list (so emit multiple InitApply when we receive an InitPage event).

Possibly, but then we are introducing buffering again inside the trampoline so that we can loop through the vector.

This does feel a bit like an optimisation, and I wonder if this, along with your other use case you are describing in your main review comment can be minimised instead with an additional reducer ala `Event::into_old_style_event_something()` that can do some bare-bones buffering for your use case?